Update LineContainer.mdx (#4518)
cpinitiative/usaco-guide: content/6_Advanced/LineContainer.mdx

So, as an overview of the algorithm:
* Sort the set of half-planes by angle; it takes $\mathcal{O}(NlogN)$ time.
* Iterate over the set. For each, perform incrementation and pop either from the front or back of the double-ended queue as necessary; it takes linear time since every half-plane can only be added or removed once.
* Once this has been completed, the convex polygon that results from the intersection can be obtained by computing the intersection points of adjacent half-planes in the double-ended queue. It will take linear time. This results in our final complexity as $\mathcal{O}(NlogN)$. In a special case where the half-planes are pre-sorted, this algorithm could be completed in $\mathcal{O}(N)$. However, for the majority of cases, we will be looking at $\mathcal{O}(NlogN)$, most of which is coming from the sorting by angle.

## Implementation

> ```cpp

Author: done

We can state that $N$ is the number of half-planes in our set. We represent lines and half-planes by a point and a vector, or in other terms, any point that lies on said line and the direction vector of the line. Each half-plane allows the region to the left of its direction vector. We can define the angle of a given half-plane as the polar coordinates of its direction vector. We can also state that a resulting intersection is always either bounded or empty, in cases where this doesn't hold, we can add 4 more half-planes which increase the size of the bounding box. For now, we shall also assume that parallel half-planes will not be present. To get a more visual representation of this, see the image below:

![halfplane1](https://github.com/cpinitiative/usaco-guide/assets/145281030/bf54b84b-038f-47bd-a7e7-2f17ed8b8504)

This half-plane has an equation $y = 2x -2$ and can be represented as $P = (1,0)$ with the direction vector $PQ$ as $(1,2)$.

> I will now describe various ways of solving this problem, explaining algorithms in increasing efficiency:

Reviewer: wait so is there a focus problem for this

Reviewer: wait i just took a look at your code
there's no main method?

Author: it's in c++?

Reviewer: no yeah but,, you know most of the code here is intended to be submitted to a problem for ac purposes
your code here has no main method, so

Author: this problem is not from any contests

Reviewer: then how can we make sure it's correct

Author: the provided solution is correct. it does work.

Reviewer: can you send a proof of ac on a problem that used it?

Author: sure, but this code does work for this problem, like official solution.
and acceptance for what test cases in what problem? there's none.
the example problem's in the module, for which this solution works.

Reviewer: > this problem

what problem?

Reviewer: which specific one?

Author: the one explained in this module with pictures

Reviewer: unless you have an actual problem with test cases i'm pretty leery of adding this

Author: here, how about this: https://www.codechef.com/problems/CHN02.
i can also add extra practice problems on this topic.

Reviewer: as long as the sol works on that

Author: proof that the solution works on this problem:
![image](https://github.com/cpinitiative/usaco-guide/assets/145281030/ea52b1c2-4830-4b1d-b577-6f3dbc57538d)

Author: also, my code does have a main method; i just looked.

Author: and once you merge this, i can add a bunch of practice problems i found (from actual contests with test cases) in the json file.

Reviewer: > also, my code does have a main method; i just looked.

what line in the file?

> and once you merge this, i can add a bunch of practice problems i found (from actual contests with test cases) in the json file.

why not add them now?

Author: bcz rn the current module is unstructured without proper sections for the problems for each topic. once this is merged, the problems section will be formatted correctly.

Reviewer: you can make the sections yourself my guy

Author: ik and i did in this pr. but idk exactly where the problems i add in the json file would show up in the current module bcz rn it has like 3 problem sections when it only needs 2 (one for half-plane, other for line-container). if this pr is merged, there will be 2 problem sections like there supposed to and then i can add the problems. look at the current module now and u'll see how disorganized it is.

Reviewer: why...would merging this turn it back into 2 problem sections?

also
> what line in the file?

Author: nvm, i guess i'll still add those problems. and nvm for that too; i thought i saw main, but nope didn't

Author: @SansPapyrus683 i added those problems for half-plane intersection in the json file in this pr #4577

	long double x, y;
	explicit Point(long double x = 0, long double y = 0) : x(x), y(y) {}

	/** @comment Operator overloads for Point operations */
};

struct Halfplane {
	Point p, pq;       /** @Point and direction vector of half-plane */
	long double angle; /** @Angle of the direction vector */

	Halfplane() {}
	Halfplane(const Point &a, const Point &b) : p(a), pq(b - a) {
		angle = atan2l(pq.y, pq.x); /** @Compute angle of direction vector */
	}

	/** @Check if a point r is outside the half-plane */
	bool out(const Point &r) { return cross(pq, r - p) < -eps; }

	/** @Operators for sorting half-planes based on angle */
};

/** @Compute the intersection of a set of half-planes */
std::vector<Point> halfplane_intersect(std::vector<Halfplane> &halfplanes) {
	Point box[4] = {Point(inf, inf), Point(-inf, inf), Point(-inf, -inf),
	                Point(inf, -inf)};
	/** @Define a bounding box as 4 half-planes */

	/** @Add bounding box halfplanes to the input set */
	for (int i = 0; i < 4; i++) {
		Halfplane aux(box[i], box[(i + 1) % 4]);
		halfplanes.push_back(aux);
	}

	/** @Sort half-planes by angle (counter-clockwise) */
	std::sort(halfplanes.begin(), halfplanes.end());

> 	/** @Remove duplicates (half-planes with same direction) */

Reviewer: uuuhhh idt these are valid doc comment things lol
https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html
see here for valid tags

Author: oh ok lol just saw. i'll fix them.

* Some of the half-planes that are located at the back of the intersection may become redundant, in which case we shall pop from the queue from the back of the double-ended queue.
* Some of the half-planes that are located at the front of the intersection may become redundant, in which case we shall pop from the queue from the back of the double-ended queue.
* The intersection may become empty from this, in which case we shall simply terminate the algorithm and return the intersection as empty.

In this case, redundant means that it doesn't change the intersection, or the half-plane could be removed and the intersection wouldn't change.

To visualize this, we shall let $H = A,B,C,D,E$. This is the set of half-planes that are currently present in the intersection. For the points, we can define them as $P = p,q,r,s$ which are intersection points of adjacent half-planes from $H$. This could be represented as such below:

> ![image2](halfplane2.png)

Reviewer 2: make sure that you import it under the images folder in 6_Advanced.
![image](https://github.com/cpinitiative/usaco-guide/assets/81596991/9afe6c80-5f46-4b47-8161-711579322b0d)

Author: the images still aren't loading for some reason even though i added them into the 6_Advanced folder and here in the files in this pr.

Author: nvm, i figured it out. the images are loading now.

## Implementation

**Time Complexity: $\mathcal{O}(N \log N)$**

<LanguageSection>
<CPPSection>

```cpp
#include <algorithm>
#include <climits>
#include <deque>
#include <vector>

const long double eps = 1e-9;
const int inf = INT32_MAX;

struct Point {
	long double x, y;
	explicit Point(long double x = 0, long double y = 0) : x(x), y(y) {}

	friend Point operator+(const Point &p, const Point &q) {
		return Point(p.x + q.x, p.y + q.y);
	}

	friend Point operator-(const Point &p, const Point &q) {
		return Point(p.x - q.x, p.y - q.y);
	}

	friend Point operator*(const Point &p, const long double &k) {
		return Point(p.x * k, p.y * k);
	}

	friend long double cross(const Point &p, const Point &q) {
		return p.x * q.y - p.y * q.x;
	}
};

struct Halfplane {
	Point p, pq;
	long double angle;

	Halfplane() {}
	Halfplane(const Point &a, const Point &b) : p(a), pq(b - a) {
		angle = atan2l(pq.y, pq.x);
	}

> 	bool out(const Point &r) { return cross(pq, r - p) < -eps; }

Reviewer: what this method do

	bool operator<(const Halfplane &e) const {
		if (fabsl(angle - e.angle) < eps) return cross(pq, e.p - p) < 0;
		return angle < e.angle;
	}

	bool operator==(const Halfplane &e) const {
		return fabsl(angle - e.angle) < eps;
	}

> 	friend Point inter(const Halfplane &s, const Halfplane &t) {

Reviewer: and this
(i mean add doc comments)
like
```cpp
/** @return something, idk lol */
```

Author: i just added doc comments in the specified format.

			len--;
		}

		// Remove invalid half-planes from the beginning of active set
		while (len > 1 && halfplanes[i].out(intersection(
		                      active_halfplanes[0], active_halfplanes[1]))) {
			active_halfplanes.pop_front();
			len--;
		}

		// Add current half-plane to active set
		active_halfplanes.push_back(halfplanes[i]);
		len++;
	}

	// Check if there are less than 3 valid half-planes
	if (len < 3) return std::vector<Point>();

	// Compute intersection points of active half-planes
	std::vector<Point> intersection_points(len);
	for (int i = 0; i < len - 1; i++) {
		intersection_points[i] =
		    intersection(active_halfplanes[i], active_halfplanes[i + 1]);
	}
	intersection_points.back() =
	    intersection(active_halfplanes[len - 1], active_halfplanes[0]);

	return intersection_points;
}

// Function to compute intersection of two half-planes
Point intersection(const Halfplane &h1, const Halfplane &h2) {
	long double den = cross(h1.pq, h2.pq);
	long double num1 = cross(h2.pq, h1.p - h2.p);

	// Calculate intersection point
	return Point(h1.p.x + (num1 / den) * h1.pq.x,
	             h1.p.y + (num1 / den) * h1.pq.y);
}

int main() {

> 	std::vector<Halfplane> halfplanes;

Reviewer: bruh
what problem does this ac on

Author: the example one

Reviewer: no most modules give an algorithm _with_ a cp-style problem

Author: so, now should i add that problem and its solution in the module as well (the one i submitted proof of ac for)?

Author: but the solution still outputs what the algorithm is supposed to do; it gives the intersection points. so a cp problem doesn't matter??? the problem explanation in the module is still a great problem.

Reviewer: the whole point of usaco guide is to give an actual problem to apply these algorithms to in both examples and exercises

Author: so should i add practice problems

Reviewer: yes
i thought this was well established 💀

Author: and should i replace that ac problem with its new solution in place of the existing one? or should i add below the existing explanation and all?

Reviewer: replace

Reviewer: oh btw can i see this new problem? it has to be an extremely straightforward application of the algorithm you describe- just like the other cses problems the guide showcases

Author: yeah, the same one i got ac for:
https://www.codechef.com/problems/CHN02

Author: also idk how to add another file to this pr, so can i add that focus problem and other practice problems in the json file in a separate pr? as in reopen that closed pr...

Reviewer: i
how do you not know
how to commit multiple files

Reviewer: but yeah the problem seems nice

Author: ok i'll figure it out.

Author: it automatically makes a different pr, i tried. so can you reopen that closed pr? then, i can also add the new ac cp-style focus problem, in addition to the practice problems i already committed there.

Reviewer: oh.
are you working on the usaco guide editor?
please just...git clone it locally

Author: no, i'm working on the web editor on github.
idk does the same thing?
so reopen that pr, things will be much easier and faster?

Reviewer: no it's very limiting
just git clone

Author: ok
edit: i tried but nope it doesn't work, still requires some merging that creates a diff pr.
so rather i'll make another pr for the json file myself, like before. i don't see how it's limiting; moreover, it's more flexible.

Reviewer: i have never heard of anyone having to make two separate prs for two separate files
are you sure you're on the right branch?

Author: yeah

Author: yeah
so, i'm merging that closed pr's branch of added problems into this pr's branch in my forked repository, but it still makes a diff pr for merging that into this. so, rather i'll make a diff pr myself to separately update it.
bcz this is taking much longer than i expected, like so close to a 100 conversations, i've never seen that number in any pr.
reopening that is much easier and definitely faster.

Reviewer: > can you tell me what exactly you're doing, in terms of commits, pushes, whatever?
maybe a transcript of what shows on your terminal would help as well

Author: uh, i just did?

Author: i'll add the ac cp-style problem now

Author: update: i added the focus problem

Reviewer: where is it

Author: right after the resources and before the explanation

Author: i added it as a focus problem like in other modules, followed by explanation and solution.

Author: i also added the practice problems

Reviewer: the main method doesn't seem to read any input from stdin

Author: i modified it. now it should work.

Author: @SansPapyrus683 is there anything else to be reviewed and modified?

Reviewer: uh overall it looks fine, but the explanation needs to be looked at and i'm a bit busy lol
sorry for the inconvenience

Author: oh it's ok

So, as an overview of the algorithm:
* Sort the set of half-planes by angle; it takes $\mathcal{O}(NlogN)$ time.
* Iterate over the set. For each, perform incrementation and pop either from the front or back of the double-ended queue as necessary; it takes linear time since every half-plane can only be added or removed once.
* Once this has been completed, the convex polygon that results from the intersection can be obtained by computing the intersection points of adjacent half-planes in the double-ended queue. It will take linear time. This results in our final complexity as $\mathcal{O}(NlogN)$. In a special case where the half-planes are pre-sorted, this algorithm could be completed in $\mathcal{O}(N)$. However, for the majority of cases, we will be looking at $\mathcal{O}(N \log N)$, most of which is coming from the sorting by angle.

## Implementation

**Time Complexity: $\mathcal{O}(N \log N)$**

<LanguageSection>
<CPPSection>

```cpp

> #include <algorithm>

Reviewer: wait i just realized
is there an oj you can submit this code to to test it or anything

Author: no, it's not from competitions/hackathons or anything. Just like an intro/explanation example problem.

Reviewer: so how can users know if their code is correct

Author: by checking the solution ???

	</Resource>
</Resources>

<FocusProblem problem="HPIsam" />

A half-plane can be defined as a planar region that consists of all points on one side of an infinite straight line, and none on the other side. We can apply computational geometry here, as an intersection can be represented as a convex polygon, where every point inside it is also inside all half-planes. If we can construct this polygon, we can retrieve the intersections.

> We can state that $N$ is the number of half-planes in our set. We represent lines and half-planes by a point and a vector, or in other terms, any point that lies on said line and the direction vector of the line. Each half-plane allows the region to the left of its direction vector. We can define the angle of a given half-plane as the polar coordinates of its direction vector. We can also state that a resulting intersection is always either bounded or empty, in cases where this doesn't hold, we can add 4 more half-planes which increase the size of the bounding box. For now, we shall also assume that parallel half-planes will not be present. To get a more visual representation of this, see the image below:

Reviewer: > We can also state that a resulting intersection is always either bounded or empty

uuuhh what

Author: maybe in complex theory, but definitely not practically.

Reviewer: it takes like 0 effort to come up with an infinite intersection

Author: ik but...

Author: doesn't show up in cp?

We can see $F$ here represented in orange. It makes the half-planes $A$ and $E$ redundant. Therefore, we can remove them from the front and back of the queue and add $F$ at the end. Then, we obtain a new intersection between $H = B,C,D,F$ with $P = q,r,t,u$. This can be shown below:

![halfplane4](https://github.com/cpinitiative/usaco-guide/assets/145281030/c2662dd6-433c-43f0-b3c4-5064e58f878e)

This is the main principle of the algorithm. However, there are a few special cases we should cover. In the case of parallel half-planes, there are 2 possible scenarios, either two that are parallel with the same direction or different directions. We should approach parallel lines with different directions the same as if we have to increase the size of the bounding box, since there will have to be at least one of the half-planes within the bounding box in between the two since they are sorted by angle. This means that the only case we have to deal with is where there are two half-planes with the same angle. To do this is straightforward: simply keep the leftmost half-plane and erase the rest since the others will eventually become redundant anyway.

So, as an overview of the algorithm:

> * Sort the set of half-planes by angle; it takes $\mathcal{O}(NlogN)$ time.

Reviewer 2: ```suggestion
* Sort the set of half-planes by angle; it takes $\mathcal{O}(N\logN)$ time.
```
use latex log

	for (int i = 0; i < 4; i++) {
		Halfplane aux(box[i], box[(i + 1) % 4]);
		halfplanes.push_back(aux);
	}

	// Sort half-planes by angle (counter-clockwise)
	std::sort(halfplanes.begin(), halfplanes.end());

	// Remove duplicates (half-planes with same direction)
	halfplanes.erase(std::unique(halfplanes.begin(), halfplanes.end()),
	                 halfplanes.end());

	std::deque<Halfplane> active_halfplanes;
	int len = 0;

	// Process each sorted half-plane
	for (int i = 0; i < int(halfplanes.size()); i++) {
		// Remove invalid half-planes from the end of active set
		while (len > 1 &&
		       halfplanes[i].out(intersection(active_halfplanes[len - 1],
		                                      active_halfplanes[len - 2]))) {
			active_halfplanes.pop_back();
			len--;
		}

		// Remove invalid half-planes from the beginning of active set
		while (len > 1 && halfplanes[i].out(intersection(
		                      active_halfplanes[0], active_halfplanes[1]))) {
			active_halfplanes.pop_front();
			len--;
		}

		// Add current half-plane to active set
		active_halfplanes.push_back(halfplanes[i]);
		len++;
	}

	// Check if there are less than 3 valid half-planes

> 	if (len < 3) return 0;

Reviewer 3: braces

Author: added

So, as an overview of the algorithm:
* Sort the set of half-planes by angle; it takes $\mathcal{O}(NlogN)$ time.
* Iterate over the set. For each, perform incrementation and pop either from the front or back of the double-ended queue as necessary; it takes linear time since every half-plane can only be added or removed once.
* Once this has been completed, the convex polygon that results from the intersection can be obtained by computing the intersection points of adjacent half-planes in the double-ended queue. It will take linear time. This results in our final complexity as $\{O}(NlogN)$. In a special case where the half-planes are pre-sorted, this algorithm could be completed in $\{O}(N)$. However, for the majority of cases, we will be looking at $\{O}(NlogN)$, most of which is coming from the sorting by angle.

> ##### Implementation

Reviewer: 5 hashtags seems a bit much...

```cpp

> const long double eps = 1e-9, inf = 1e9;

Reviewer: just use INT32_MAX for inf

struct Point {

> 

Reviewer: ```suggestion
```

	long double x, y;
	explicit Point(long double x = 0, long double y = 0) : x(x), y(y) {}

	friend Point operator+(const Point &p, const Point &q) {
		return Point(p.x + q.x, p.y + q.y);
	}

	friend Point operator-(const Point &p, const Point &q) {
		return Point(p.x - q.x, p.y - q.y);
	}

	friend Point operator*(const Point &p, const long double &k) {
		return Point(p.x * k, p.y * k);
	}

	friend long double cross(const Point &p, const Point &q) {
		return p.x * q.y - p.y * q.x;
	}
};

struct Halfplane {

> 

Reviewer: ```suggestion
```

	Point p, pq;
	long double angle;

	Halfplane() {}
	Halfplane(const Point &a, const Point &b) : p(a), pq(b - a) {
		angle = atan2l(pq.y, pq.x);
	}

	bool out(const Point &r) { return cross(pq, r - p) < -eps; }

	bool operator<(const Halfplane &e) const {
		if (fabsl(angle - e.angle) < eps) return cross(pq, e.p - p) < 0;
		return angle < e.angle;
	}

	bool operator==(const Halfplane &e) const {
		return fabsl(angle - e.angle) < eps;
	}

	friend Point inter(const Halfplane &s, const Halfplane &t) {
		long double alpha = cross((t.p - s.p), t.pq) / cross(s.pq, t.pq);
		return s.p + (s.pq * alpha);
	}
};

// Actual algorithm

> vector<Point> hp_intersect(vector<Halfplane> &H) {

Reviewer: 1. doc comment explaining what this does
2. rename `H` to something more informative or at least `h` bc snake_case

	Point box[4] = {Point(inf, inf), Point(-inf, inf), Point(-inf, -inf),
	                Point(inf, -inf)};

	for (int i = 0; i < 4; i++) {
		Halfplane aux(box[i], box[(i + 1) % 4]);
		H.push_back(aux);
	}

	sort(H.begin(), H.end());
	H.erase(unique(H.begin(), H.end()), H.end());

	deque<Halfplane> dq;
	int len = 0;
	for (int i = 0; i < int(H.size()); i++) {

> 

Reviewer: ```suggestion
```

		while (len > 1 && H[i].out(inter(dq[len - 1], dq[len - 2]))) {
			dq.pop_back();

> 			--len;

Reviewer: can you use postincrements instead of preincrements everywhere

So, as an overview of the algorithm:
* Sort the set of half-planes by angle; it takes $\mathcal{O}(NlogN)$ time.
* Iterate over the set. For each, perform incrementation and pop either from the front or back of the double-ended queue as necessary; it takes linear time since every half-plane can only be added or removed once.
* Once this has been completed, the convex polygon that results from the intersection can be obtained by computing the intersection points of adjacent half-planes in the double-ended queue. It will take linear time. This results in our final complexity as $\{O}(NlogN)$. In a special case where the half-planes are pre-sorted, this algorithm could be completed in $\{O}(N)$. However, for the majority of cases, we will be looking at $\{O}(NlogN)$, most of which is coming from the sorting by angle.

#### Implementation

```cpp
#include <climits>
#include <deque>
#include <vector>
#include <algorithm>

const long double eps = 1e-9;
const int inf = INT32_MAX;

struct Point {
    long double x, y;

>     explicit Point(long double x = 0, long double y = 0) : x(x), y(y) {}

Reviewer: hey man you know the precommit bot is changing things to tabs for a reason right

Author: so, what do u want me to change?

Reviewer: no like i just find it interesting how pre commit changed everything to tabs and yet you always change it back to spaces

const long double eps = 1e-9;
const int inf = INT32_MAX;

struct Point {
	long double x, y;
	explicit Point(long double x = 0, long double y = 0) : x(x), y(y) {}

	friend Point operator+(const Point &p, const Point &q) {
		return Point(p.x + q.x, p.y + q.y);
	}

	friend Point operator-(const Point &p, const Point &q) {
		return Point(p.x - q.x, p.y - q.y);
	}

	friend Point operator*(const Point &p, const long double &k) {
		return Point(p.x * k, p.y * k);
	}

	friend long double cross(const Point &p, const Point &q) {
		return p.x * q.y - p.y * q.x;
	}
};

struct Halfplane {
	Point p, pq;
	long double angle;

	Halfplane() {}
	Halfplane(const Point &a, const Point &b) : p(a), pq(b - a) {
		angle = atan2l(pq.y, pq.x);
	}

	bool out(const Point &r) { return cross(pq, r - p) < -eps; }

	bool operator<(const Halfplane &e) const {
		if (fabsl(angle - e.angle) < eps) return cross(pq, e.p - p) < 0;
		return angle < e.angle;
	}

	bool operator==(const Halfplane &e) const {
		return fabsl(angle - e.angle) < eps;
	}

	friend Point inter(const Halfplane &s, const Halfplane &t) {
		long double alpha = cross((t.p - s.p), t.pq) / cross(s.pq, t.pq);
		return s.p + (s.pq * alpha);
	}
};

> // Actual algorithm: Computes the intersection of a set of halfplanes

Reviewer: ```suggestion
/** @return the intersection of a set of halfplanes */
```

	</Resource>
</Resources>

A half-plane can be defined as a planar region that consists of all points on one side of an infinite straight line, and none on the other side. We can apply computational geometry here, as an intersection can be represented as a convex polygon, where every point inside it is also inside all half-planes. If we can construct this polygon, we can retrieve the intersections.

### Example Problem

> 

Reviewer: if there's only 1 example problem, no need for this header
also the rest of the module doesn't follow the format i gave

Author: I removed the header since there's just one focus problem, and I also have the explanation and implementation sections, as detailed in the document. If you have any more suggestions, please specify them (bcz I think the module looks good now).
